In [3]:
import numpy as np
import pandas as pd
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers, models

In [139]:
data = pd.read_csv("Train_file(7).csv",header=None)
data.head()

0     1     2     3      4      5      6      7      8      9   ...  \
0 -10.10  2.52  1.82  1.51 -1.720 -1.890 -1.110 -1.880  1.280 -1.350  ...   
1  -4.89 -5.69 -3.36 -2.13  0.261 -1.500  0.754 -1.680  0.495 -1.080  ...   
2  -1.69 -5.09 -4.16 -2.41 -4.750 -0.731 -0.726 -2.310  0.667  0.517  ...   
3  -4.24 -5.01 -3.35 -3.91  1.230  1.220 -0.399 -1.620 -0.428  0.645  ...   
4  -3.10 -6.13 -3.25 -3.05  0.591  0.893 -1.930 -0.913  0.228 -0.978  ...   

        51      52       53      54      55      56      57      58      59  \
0  0.02830 -0.1630  0.03300  0.0414  0.0622 -0.0254  0.0721 -0.0541  0.0498   
1  0.00944 -0.0259 -0.10700 -0.0554  0.0102 -0.0977 -0.1050 -0.0885 -0.0220   
2  0.05000 -0.0557  0.00167 -0.2040  0.2260  0.2340  0.1100  0.0360 -0.2010   
3 -0.05950  0.0557 -0.00279 -0.0288 -0.0819 -0.0958  0.0931 -0.0223  0.1490   
4  0.12800 -0.2090 -0.07580  0.1210 -0.0852  0.0346  0.2340  0.0280  0.1970   

    60  
0  1.0  
1  0.0  
2  0.0  
3  0.0  
4  1.0  

[5 rows x 61 columns]

In [208]:
def pca(X,k):
     
    X_centered = X - np.mean(X, axis=0)
    covariance_matrix = np.cov(np.transpose(X_centered))
    print(covariance_matrix.shape)

    eigenvalues, eigenvectors = np.linalg.eig(covariance_matrix)
    
    sorted_indices = np.argsort(eigenvalues)[::-1]
    sorted_eigenvalues = eigenvalues[sorted_indices]
    sorted_eigenvectors = eigenvectors[:,sorted_indices]
    k_eigenvectors = sorted_eigenvectors[:,:k]
    print(np.linalg.norm(sorted_eigenvectors[:,1]))
    print(k_eigenvectors.shape)
    reduced_dataset = np.dot(X_centered, k_eigenvectors)
    
    print(reduced_dataset.shape)
    return reduced_dataset,train_mean

In [203]:
def pca_test(X,k,train_mean):
    print(train_mean)
    X_centered = X - train_mean
    covariance_matrix = np.cov(np.transpose(X_centered))
    print(covariance_matrix.shape)

    eigenvalues, eigenvectors = np.linalg.eig(covariance_matrix)
    
    sorted_indices = np.argsort(eigenvalues)[::-1]
    sorted_eigenvalues = eigenvalues[sorted_indices]
    sorted_eigenvectors = eigenvectors[:,sorted_indices]
    k_eigenvectors = sorted_eigenvectors[:,:k]
    print(k_eigenvectors.shape)
    reduced_dataset = np.dot(X_centered, k_eigenvectors)
    
    print(reduced_dataset.shape)
    return reduced_dataset 

In [209]:
data = np.array(data)
train_data = data[:,:-1]
train_labels = data[:,-1]
data_new,train_mean  = pca(train_data,20)

(60, 60)
0.9999999999999999
(60, 20)
(672, 20)


In [132]:
model1 = keras.Sequential(
    [
        layers.Dense(10, activation="relu", name="layer1"),
        layers.Dense(1, activation="sigmoid")
    ]
)

In [133]:
model1.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=1e-3),
              loss=tf.keras.losses.BinaryCrossentropy(),metrics=["acc"])

In [134]:
model1.fit(data_new, train_labels, epochs=10)

Epoch 1/100
21/21 [==============================] - 1s 2ms/step - loss: 1.2122 - acc: 0.3289
Epoch 2/100
21/21 [==============================] - 0s 3ms/step - loss: 1.0899 - acc: 0.3839
Epoch 3/100
21/21 [==============================] - 0s 5ms/step - loss: 0.9818 - acc: 0.4435
Epoch 4/100
21/21 [==============================] - 0s 3ms/step - loss: 0.8867 - acc: 0.4970
Epoch 5/100
21/21 [==============================] - 0s 4ms/step - loss: 0.8050 - acc: 0.5580
Epoch 6/100
21/21 [==============================] - 0s 3ms/step - loss: 0.7316 - acc: 0.6057
Epoch 7/100
21/21 [==============================] - 0s 3ms/step - loss: 0.6647 - acc: 0.6652
Epoch 8/100
21/21 [==============================] - 0s 3ms/step - loss: 0.6070 - acc: 0.6875
Epoch 9/100
21/21 [==============================] - 0s 2ms/step - loss: 0.5540 - acc: 0.7262
Epoch 10/100
21/21 [==============================] - 0s 2ms/step - loss: 0.5064 - acc: 0.7530
Epoch 11/100
21/21 [==============================] - 0s 2m

In [154]:
model2 = keras.Sequential(
    [
        layers.Dense(40, activation="relu", name="layer1"),
        layers.Dense(20, activation="relu", name="layer2"),
        layers.Dense(10, activation="relu", name="layer3"),
        layers.Dense(1, activation="sigmoid")
    ]
)

In [155]:
model2.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=1e-3),
              loss=tf.keras.losses.BinaryCrossentropy(),metrics=["acc"])

In [156]:
model2.fit(train_data, train_labels, epochs=12)

Epoch 1/12
21/21 [==============================] - 1s 2ms/step - loss: 0.6173 - acc: 0.6310
Epoch 2/12
21/21 [==============================] - 0s 2ms/step - loss: 0.4407 - acc: 0.8452
Epoch 3/12
21/21 [==============================] - 0s 2ms/step - loss: 0.3436 - acc: 0.8810
Epoch 4/12
21/21 [==============================] - 0s 2ms/step - loss: 0.2754 - acc: 0.9226
Epoch 5/12
21/21 [==============================] - 0s 2ms/step - loss: 0.2361 - acc: 0.9152
Epoch 6/12
21/21 [==============================] - 0s 2ms/step - loss: 0.2033 - acc: 0.9256
Epoch 7/12
21/21 [==============================] - 0s 2ms/step - loss: 0.1817 - acc: 0.9315
Epoch 8/12
21/21 [==============================] - 0s 2ms/step - loss: 0.1655 - acc: 0.9405
Epoch 9/12
21/21 [==============================] - 0s 2ms/step - loss: 0.1488 - acc: 0.9449
Epoch 10/12
21/21 [==============================] - 0s 2ms/step - loss: 0.1354 - acc: 0.9494
Epoch 11/12
21/21 [==============================] - 0s 2ms/step - lo

In [194]:
data1 = pd.read_csv("Test_file(3).csv",header=None)
data1.head()

0     1       2       3     4      5     6      7      8       9   ...  \
0   3.86 -4.78 -3.9100 -0.5360 -1.55 -1.500 -0.86  0.372 -0.649 -0.3920  ...   
1 -10.10 -5.69 -0.0454 -1.8100  3.86 -0.612 -1.35 -0.211  0.593 -0.0933  ...   
2  -1.75 -2.70  1.5100  0.0815 -1.60 -1.850 -2.17 -3.240 -0.541  0.4420  ...   
3  -3.17 -4.74  3.3600 -2.2000 -3.64 -0.942 -3.09  0.144 -1.070  0.4260  ...   
4  -3.52 -3.66  3.4100  0.6650 -5.04 -3.110 -1.92 -0.519  0.085 -2.4100  ...   

        51      52      53      54      55      56       57       58      59  \
0  0.00905  0.0581 -0.0947  0.0219  0.0424 -0.0362  0.01170 -0.00263 -0.0132   
1  0.10100  0.0159 -0.0465  0.0828 -0.0379  0.0256 -0.17400  0.01220 -0.0448   
2 -0.07030 -0.0123  0.0526 -0.0222 -0.0239  0.1590  0.00521  0.01560 -0.0216   
3  0.19400 -0.0504  0.1200 -0.1580 -0.0650 -0.1650 -0.17300 -0.09610  0.0643   
4  0.03590 -0.0455 -0.1060  0.0176  0.0218  0.0730  0.02720  0.02510  0.0244   

    60  
0  1.0  
1  0.0  
2  1.0  
3  1.0  
4  0.0  

[5 rows x 61 columns]

In [196]:
data1  = np.array(data1)
test_data = data1[:,:-1]
test_labels = data1[:,-1]

In [159]:
test_scores = model2.evaluate(test_data, test_labels, verbose=2)

21/21 - 0s - loss: 0.2028 - acc: 0.9345 - 172ms/epoch - 8ms/step


In [202]:
print(train_mean)

[-3.56950924e+00 -4.05890878e+00 -1.19494210e+00 -2.47403405e+00
 -1.32513903e+00 -5.39016146e-01 -1.60181624e+00 -6.68561205e-01
 -4.04297798e-01 -2.84676301e-01 -4.06978649e-01 -5.17359702e-01
 -9.63399509e-02 -6.28155060e-02 -7.52041964e-02  7.11456801e-02
 -5.82781845e-02 -2.04224568e-01  2.38735045e-02  3.02044940e-02
  4.70101399e-02  4.41390506e-02  1.89229390e-02  3.18245402e-03
  7.22147619e-03 -2.59076935e-03  1.32611872e-02 -1.07867247e-02
  5.04873810e-03 -1.05709452e-02  2.41312946e-03  1.52804814e-02
 -9.26935565e-03 -5.01051339e-03 -3.87188839e-03  2.81425937e-03
  2.39322619e-03 -9.82080952e-03 -1.15771280e-03 -8.61211905e-04
  8.01491667e-03  3.11105818e-03 -2.13408036e-03  2.63398318e-03
 -1.68754137e-03  3.77139435e-03  1.81021577e-03 -4.31814137e-03
  2.63550744e-03 -3.40152515e-03 -3.53383929e-03 -5.07733740e-04
  3.09517857e-04  3.53776637e-04 -4.84585417e-03 -5.83252976e-05
 -6.28090923e-03 -2.36755952e-06  1.30838512e-03  6.25818155e-04]


In [204]:
test_data_new  = pca_test(test_data,20,train_mean)

[-3.56950924e+00 -4.05890878e+00 -1.19494210e+00 -2.47403405e+00
 -1.32513903e+00 -5.39016146e-01 -1.60181624e+00 -6.68561205e-01
 -4.04297798e-01 -2.84676301e-01 -4.06978649e-01 -5.17359702e-01
 -9.63399509e-02 -6.28155060e-02 -7.52041964e-02  7.11456801e-02
 -5.82781845e-02 -2.04224568e-01  2.38735045e-02  3.02044940e-02
  4.70101399e-02  4.41390506e-02  1.89229390e-02  3.18245402e-03
  7.22147619e-03 -2.59076935e-03  1.32611872e-02 -1.07867247e-02
  5.04873810e-03 -1.05709452e-02  2.41312946e-03  1.52804814e-02
 -9.26935565e-03 -5.01051339e-03 -3.87188839e-03  2.81425937e-03
  2.39322619e-03 -9.82080952e-03 -1.15771280e-03 -8.61211905e-04
  8.01491667e-03  3.11105818e-03 -2.13408036e-03  2.63398318e-03
 -1.68754137e-03  3.77139435e-03  1.81021577e-03 -4.31814137e-03
  2.63550744e-03 -3.40152515e-03 -3.53383929e-03 -5.07733740e-04
  3.09517857e-04  3.53776637e-04 -4.84585417e-03 -5.83252976e-05
 -6.28090923e-03 -2.36755952e-06  1.30838512e-03  6.25818155e-04]
(60, 60)
(60, 20)
(672, 

In [205]:
test_scores = model1.evaluate(test_data_new, test_labels, verbose=2)

21/21 - 0s - loss: 4.3261 - acc: 0.1860 - 68ms/epoch - 3ms/step


Model 1: I have converted (672,60) dimentional Training set to (672,20) dimension
 i.e. features are reduced 20. Then passed to MLP with 10 hidden layer 
 and then used output as sigmoid, I am getting less accuracy than model 2, which is only MLP model.

Model 2:   MLP with 40,20,10 hidden layer and then used output as sigmoid, giving test accuracy 93%.